### PR TITLE
Introduce HashedPostStateProvider

### PR DESCRIPTION
**Pull request by frisitano**
# Overivew
This PR introduces the `HashedPostStateProvider` which is used to hash `BundleState` using the `KeyHasher` from the `StateCommitment` object.

### CODE REVIEW

File: crates/stages/stages/src/stages/execution.rs [hunk 1]
```diff
@@ -262,8 +262,8 @@ where
         + StaticFileProviderFactory
         + StatsReader
         + BlockHashReader
-        + StateWriter<Receipt = reth_primitives::Receipt>
-        + StateCommitmentProvider,
+        + StateCommitmentProvider
+        + StateWriter<Receipt = reth_primitives::Receipt>,
```
Comment (greged93) [inline crates/stages/stages/src/stages/execution.rs]: nit: should we remove this? It adds unnecessary diff with upstream

File: crates/storage/provider/src/providers/bundle_state_provider.rs [hunk 2]
```diff
@@ -182,11 +182,19 @@ impl<SP: StateProvider, EDP: ExecutionDataProvider> StateProofProvider
...
+        input.prepend(self.hashed_post_state(bundle_state));
         self.state_provider.witness(input, target)
     }
 }
 
+impl<SP: StateProvider, EDP: ExecutionDataProvider> HashedPostStateProvider
+    for BundleStateProvider<SP, EDP>
+{
+    fn hashed_post_state(&self, bundle_state: &revm::db::BundleState) -> HashedPostState {
+        self.state_provider.hashed_post_state(bundle_state)
```
Comment (greged93) [inline crates/storage/provider/src/providers/bundle_state_provider.rs:194]: Should this be
```rust
        let mut state = self.block_execution_data_provider.execution_outcome().state();
        state.extend(bundle_state);
        self.state_provider.hashed_post_state(state)
```
?

**Review by greged93 — COMMENTED** [review]
Just one question and one nit!

File: crates/stages/stages/src/stages/execution.rs (hunk 1 above)
Comment (frisitano) [reply crates/stages/stages/src/stages/execution.rs]: Do you mean to reverse the ordering? If so yes we should.

Comment (greged93) [reply crates/stages/stages/src/stages/execution.rs]: yes reverse the order, I meant remove the diff

File: crates/storage/provider/src/providers/bundle_state_provider.rs (hunk 2 above)
Comment (frisitano) [reply crates/storage/provider/src/providers/bundle_state_provider.rs:194]: Good catch, yes I think this makes most sense.

Comment (frisitano) [reply crates/storage/provider/src/providers/bundle_state_provider.rs:194]: I've looked into this a bit more and whilst I agree that it logically makes sense for the `BundleStateProvider` to include the latest state from the `BundleStateProvider` it introduces some complexities when the caller wants to chain methods from the provider downstream. I.e. 
```rust
let hashed_state = provider.hashed_post_state(bundle);
let state_root = provider.state_root(hashed_state);
```

I think it's easier to reason if we assume that the `HashedPostStateProvider` is stateless for the `BundleStateProvider`. Which is the pattern we use for all other providers that implement `HashedPostStateProvider`. It also helps if we stay algined with upstream here https://github.com/paradigmxyz/reth/pull/12607.

File: crates/stages/stages/src/stages/execution.rs (hunk 1 above)
Comment (frisitano) [reply crates/stages/stages/src/stages/execution.rs]: done

File: crates/storage/provider/src/providers/bundle_state_provider.rs (hunk 2 above)
Comment (greged93) [reply crates/storage/provider/src/providers/bundle_state_provider.rs:194]: > let hashed_state = provider.hashed_post_state(bundle);
let state_root = provider.state_root(hashed_state);

I see, this does complicate things.

**Review by greged93 — APPROVED** [review]
LGTM!
